runner/tasks.py

```diff
@@ -14,55 +14,88 @@
 logger = logging.getLogger(__name__)
 
 
+def create_jobs_from_operator(operator):
+    jobs = operator.get_jobs()
+
+    valid_jobs, invalid_jobs = [], []
+    for job in jobs:
+        valid_jobs.append(job) if job[0].is_valid() else invalid_jobs.append(job)
+
+    operator_run = OperatorRun.objects.create(operator=operator, total_num_runs=len(valid_jobs))
+    Run.objects.bulk_update(valid_jobs, operator_run_id=operator_run.id)
+
+    for job in valid_jobs:
+        logger.info("Creating Run object")
+        run = job[0].save()
+        logger.info("Run object created with id: %s" % str(run.id))
+        create_run_task.delay(str(run.id), job[1], None)
+
+    for job in invalid_jobs:
+        logger.error("Job invalid: %s" % str(job[0].errors))
 
 @shared_task
 def create_jobs_from_request(request_id, operator_id):
     logger.info("Creating operator id %s for requestId: %s" % (operator_id, request_id))
-    operator_info = Operator.objects.get(id=operator_id)
-    operator = OperatorFactory.get_by_class_name(operator_info.class_name, request_id)
+    operator_model = Operator.objects.get(id=operator_id)
+    operator = OperatorFactory.get_by_model(operator_model, request_id=request_id)
+    create_jobs_from_operator(operator)
 
-    jobs = operator.get_jobs()
 
-    for job in jobs:
-        if job[0].is_valid():
-            logger.info("Creating Run object")
-            run = job[0].save()
-            logger.info("Run object created with id: %s" % str(run.id))
-            create_run_task.delay(str(run.id), job[1], None)
-        else:
-            logger.error("Job invalid: %s" % str(job[0].errors))
+@shared_task
+def create_jobs_from_chaining(to_operator_id, from_operator_id, run_ids=[]):
+    logger.info("Creating operator id %s from chaining: %s" % (operator_id, from_operator_id))
+    operator_model = Operator.objects.get(id=to_operator_id)
+    operator = OperatorFactory.get_by_model(operator_model, run_ids=run_ids)
+    create_jobs_from_operator(operator)
 
 
-# TODO remove this when the Celery backlog is flushed and OperatorViewSet is updated
 @shared_task
-def operator_job(request_id, pipeline_type):
-    # warnings.warn("operator_job to be deprecated for create_jobs_from_request", DeprecationWarning, stacklevel=2)
-    logger.info("Creating operator %s for requestId: %s" % (pipeline_type, request_id))
-    operator = OperatorFactory.factory(pipeline_type, request_id)
-    jobs = operator.get_jobs()
-    for job in jobs:
-        if job[0].is_valid():
-            logger.info("Creating Run object")
-            run = job[0].save()
-            logger.info("Run object created with id: %s" % str(run.id))
-            create_run_task.delay(str(run.id), job[1], None)
-        else:
-            logger.error("Job invalid: %s" % str(job[0].errors))
+def process_triggers():
```

Comment:
as discussed, this shouldn't be a celery beat, but rather, a single operator run should trigger its dependents when its status is changed to complete. I'm not sure where we should update its status to complete since there are some race conditions at play, so I left this here for now.

Reply:
Sure.